### PR TITLE
[XPU][FP8] Add block-scaled MoE prefill GEMM

### PR DESCRIPTION
**Pull request by gc-fu**
## Summary

- route block-scaled grouped MoE calls with at least 128 routed rows to a large-M prefill kernel
- build compact 64-row expert-tile descriptors on device without a device-to-host synchronization
- use FP16 DPAS after E4M3 block dequantization for the prefill path while preserving the existing decode path
- add a pytest-collectable correctness and microbenchmark test for the prefill kernel

## Scope

This is the next kernel-only stage after #565. It does not change the public operator schema or require a framework change; the existing `esimd_moe_gemm_fp8_blockscale` dispatch selects the prefill implementation internally for `total_tokens >= 128`.

The later descriptor-buffer reuse optimization is intentionally not included in this PR and will remain a separate stage.

## Testing

Not run as part of PR creation.

### CODE REVIEW

**Comment by gc-fu** [comment]
Validation completed on XPU 0,1 with framework `4d20a859d`, kernel `465af0d`, and oneCCL `/opt/intel/oneapi/ccl/2021.15/lib/libccl.so.1`.

Focused large-M MoE prefill test:
- pytest: `1 passed`
- mean relative error: `3.375048e-4`
- cosine similarity: `1.0`
- microbenchmark: `0.0214 ms`

GSM8K per-block accuracy:
- 30 x 100 questions, all rounds PASS
- mean `97.567%`, range `96%-99%`
- invalid `0`, NaN `0`, device errors `0`, all health checks PASS

The mean is 1.058 percentage points below the Stage 3 24-round result (`98.625%`). Both runs pass, but the differing round counts and numerical shift are noted for review.

32k input / 2k output performance, one measured round, reusing the accuracy-warmed service:

| Batch | Output tok/s | TTFT (ms) | TPOT (ms) | E2E (ms) |
|---:|---:|---:|---:|---:|
| 1 | 51.5918 | 5438.121 | 16.736 | 39695.771 |
| 4 | 116.8256 | 10816.796 | 28.904 | 69983.526 |

Compared with Stage 3 under the same service-reuse policy:
- Batch 1: throughput `+36.08%`, TTFT `-73.03%`, E2E `-26.51%`.
- Batch 4: throughput `+89.89%`, TTFT `-79.44%`, TPOT `-26.42%`, E2E `-47.39%`.

Overall validation status: PASS.

File: vllm/custom-esimd-kernels-vllm/tests/test_fp8_blockscale_moe_prefill.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+import custom_esimd_kernels_vllm as esimd
+
+sys.path.insert(0, str(Path(__file__).resolve().parent))
+from test_moe_forward_full_fp8_block import dequant, quantize_block
+
+
+def test_fp8_blockscale_moe_prefill() -> None:
+    torch.manual_seed(20260713)
+    device = "xpu"
```
[the remark this answers is not in the document]
Comment (gc-fu) [reply vllm/custom-esimd-kernels-vllm/tests/test_fp8_blockscale_moe_prefill.py]: Fixed in ca6888b. The pytest correctness path now skips when XPU is unavailable and no longer runs the warmup/timed benchmark loop; direct script execution retains the microbenchmark.

File: vllm/custom-esimd-kernels-vllm/tests/test_fp8_blockscale_moe_prefill.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    torch.xpu.synchronize()
+    begin = time.perf_counter()
+    for _ in range(100):
+        esimd.esimd_moe_gemm_fp8_blockscale(
+            x, weight, scale, out, offsets, n, k, experts
+        )
+    torch.xpu.synchronize()
+    mean_ms = (time.perf_counter() - begin) * 10.0
+    print(f"mean_rel={mean_rel:.6e} cosine={cosine:.8f} mean_ms={mean_ms:.4f}")
+    print("RESULT: ALL OK")
```
[the remark this answers is not in the document]
Comment (gc-fu) [reply vllm/custom-esimd-kernels-vllm/tests/test_fp8_blockscale_moe_prefill.py]: Fixed in ca6888b. Correctness remains the default pytest workload, while the warmup and 100 timed iterations only run through the explicit script entry point.

File: vllm/custom-esimd-kernels-vllm/tests/test_fp8_blockscale_moe_prefill.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import time
+import sys
+from pathlib import Path
+
+import torch
+
+import custom_esimd_kernels_vllm as esimd
+
+sys.path.insert(0, str(Path(__file__).resolve().parent))
+from test_moe_forward_full_fp8_block import dequant, quantize_block
```
[the remark this answers is not in the document]
Comment (gc-fu) [reply vllm/custom-esimd-kernels-vllm/tests/test_fp8_blockscale_moe_prefill.py]: Fixed in ca6888b. The test is now self-contained and no longer mutates sys.path or imports another test module.

File: vllm/custom-esimd-kernels-vllm/csrc/xpu/esimd_kernels/fp8_moe_gemm_blockscale.h [hunk 4]
```diff
@@ -134,6 +171,156 @@ struct moe_gemv_block_kernel {
...
+  const int Kb = K / block_k;
+  const int n_tiles = (N + 15) / 16;
+  const int wg_cap = (K <= 512) ? 51200 : 38400;
+  int n_per_wg = 1;
+  int total_wgs = tile_capacity * n_tiles;
+  while (total_wgs > wg_cap && n_per_wg < n_tiles) {
+    n_per_wg++;
+    while (n_per_wg < n_tiles && n_tiles % n_per_wg != 0) n_per_wg++;
+    total_wgs = tile_capacity * ((n_tiles + n_per_wg - 1) / n_per_wg);
+  }
```
[the remark this answers is not in the document]
Comment (gc-fu) [reply vllm/custom-esimd-kernels-vllm/csrc/xpu/esimd_kernels/fp8_moe_gemm_blockscale.h:315]: Fixed in ca6888b. The code now documents that these are empirical BMG grid-shaping caps, why the cap differs by K, and that coalescing N tiles does not truncate computation.
